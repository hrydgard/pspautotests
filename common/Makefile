TARGET_LIB = libcommon.a
OBJS = common.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak
