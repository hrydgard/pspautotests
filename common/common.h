#ifndef _COMMON_H
#define _COMMON_H 1

#ifdef __cplusplus
extern "C" {
#endif

// Just to make IntelliSense happy, not attempting to compile.
#ifdef _MSC_VER
#define __builtin_va_list int
#include <stdarg.h>

#define __inline__ inline
#define __attribute__(...)
#define __extension__

#ifndef __builtin_va_start
void __va_start(va_list, ...);
void *__va_arg(va_list, ...);
void __va_end(va_list);

#define __builtin_va_start __va_start
#define __builtin_va_arg __va_arg
#define __builtin_va_end __va_end
#endif
#endif

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <pspctrl.h>
#include <pspkerneltypes.h>

#undef main
#define main test_main

extern unsigned int RUNNING_ON_EMULATOR;
extern unsigned int CHECKPOINT_ENABLE_TIME;
// Causes rescheduling (sceIoWrite) but easier to debug in the emulator.
extern unsigned int CHECKPOINT_OUTPUT_DIRECT;
extern char schedfBuffer[65536];
extern unsigned int schedfBufferPos;

void schedf(const char *format, ...);
void flushschedf();
int reschedFunc(SceSize argc, void *argp);
void checkpoint(const char *format, ...);
void checkpointNext(const char *title);

#define ARRAY_SIZE(a) (sizeof((a)) / (sizeof((a)[0])))

void emulatorEmitScreenshot();
void emulatorSendSceCtrlData(SceCtrlData* pad_data);

/*
void emitInt(int v);
void emitFloat(float v);
void emitString(char *v);
void emitComment(char *v);
void emitMemoryBlock(void *address, unsigned int size);
void emitHex(void *address, unsigned int size);
#define emitStringf(format, ...) { char temp[1024]; sprintf(temp, format, __VA_ARGS__); emitString(temp); }
*/

#ifdef __cplusplus
}
#endif

#endif