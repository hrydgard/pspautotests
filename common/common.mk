# Include this to simplify your test include file.
#
# Usage:
#
# TARGETS=mytest
# include ../../path/to/common/common.mk

BUILD_PRX = 1
USE_PSPSDK_LIBC = 1
PSP_FW_VERSION = 500

INCDIR := $(INCDIR) . $(COMMON_DIR)
LIBDIR := $(LIBDIR) . $(COMMON_DIR)

ifndef CFLAGS
CFLAGS = -g -G0 -Wall -O0 -fno-strict-aliasing
endif
ifndef CXXFLAGS
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti 
endif
ifndef ASFLAGS
ASFLAGS = $(CFLAGS)
endif
ifndef LDFLAGS
LDFLAGS = -G0
endif

ifndef LIBS
LIBS = -lpspgu -lpsprtc -lpspctrl -lpspmath -lcommon -lc -lm
endif

TARGET = $(firstword $(TARGETS))
OBJS = $(firstword $(TARGETS)).o

PSPSDK = $(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.elf: %.o $(EXTRA_OBJS) $(EXPORT_OBJ)
	$(LINK.c) $^ $(LIBS) -o $@
	$(FIXUP) $@

%.prx: %.elf
	psp-prxgen $< $@

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

all: $(TARGETS:=.prx)
clean: EXTRA_TARGETS:=$(EXTRA_TARGETS) $(TARGETS:=.prx)
