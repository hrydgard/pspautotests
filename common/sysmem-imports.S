.set noreorder

#include "pspimport.s"

IMPORT_START "SysMemUserForUser",0x40000000

IMPORT_FUNC  "SysMemUserForUser",0xA291F107,sceKernelMaxFreeMemSize
IMPORT_FUNC  "SysMemUserForUser",0xF919F628,sceKernelTotalFreeMemSize
IMPORT_FUNC  "SysMemUserForUser",0x2A3E5280,sceKernelQueryMemoryInfo

IMPORT_FUNC  "SysMemUserForUser",0x3FC9AE6A,sceKernelDevkitVersion
IMPORT_FUNC  "SysMemUserForUser",0x13A5ABEF,sceKernelPrintf
IMPORT_FUNC  "SysMemUserForUser",0x39F49610,sceKernelGetPTRIG
IMPORT_FUNC  "SysMemUserForUser",0x6231A71D,sceKernelSetPTRIG

IMPORT_FUNC  "SysMemUserForUser",0x237DBD4F,sceKernelAllocPartitionMemory
IMPORT_FUNC  "SysMemUserForUser",0xB6D61D02,sceKernelFreePartitionMemory
IMPORT_FUNC  "SysMemUserForUser",0x9D9A5BA1,sceKernelGetBlockHeadAddr

IMPORT_FUNC  "SysMemUserForUser",0xDB83A952,sceKernelGetMemoryBlockPtr
IMPORT_FUNC  "SysMemUserForUser",0xFE707FDF,sceKernelAllocMemoryBlock
IMPORT_FUNC  "SysMemUserForUser",0x50F61D8A,sceKernelFreeMemoryBlock

IMPORT_FUNC  "SysMemUserForUser",0xF77D77CB,sceKernelSetCompilerVersion
IMPORT_FUNC  "SysMemUserForUser",0xFC114573,sceKernelGetCompiledSdkVersion
IMPORT_FUNC  "SysMemUserForUser",0x7591C7DB,sceKernelSetCompiledSdkVersion
IMPORT_FUNC  "SysMemUserForUser",0x342061E5,sceKernelSetCompiledSdkVersion370
IMPORT_FUNC  "SysMemUserForUser",0x315AD3A0,sceKernelSetCompiledSdkVersion380_390
IMPORT_FUNC  "SysMemUserForUser",0xEBD5C3E6,sceKernelSetCompiledSdkVersion395
IMPORT_FUNC  "SysMemUserForUser",0x057E7380,sceKernelSetCompiledSdkVersion401_402
IMPORT_FUNC  "SysMemUserForUser",0x91DE343C,sceKernelSetCompiledSdkVersion500_505
IMPORT_FUNC  "SysMemUserForUser",0x7893F79A,sceKernelSetCompiledSdkVersion507
IMPORT_FUNC  "SysMemUserForUser",0x35669D4C,sceKernelSetCompiledSdkVersion600_602
IMPORT_FUNC  "SysMemUserForUser",0x1B4217BC,sceKernelSetCompiledSdkVersion603_605
IMPORT_FUNC  "SysMemUserForUser",0x358CA1BB,sceKernelSetCompiledSdkVersion606

# Not sure if correct.
IMPORT_FUNC  "SysMemUserForUser",0xA6848DF8,sceKernelSetUsersystemLibWork
IMPORT_FUNC  "SysMemUserForUser",0xACBD88CA,sceKernelTotalMemSize
IMPORT_FUNC  "SysMemUserForUser",0xD8DE5C1E,SysMemUserForUser_D8DE5C1E
IMPORT_FUNC  "SysMemUserForUser",0x945E45DA,SysMemUserForUser_945E45DA