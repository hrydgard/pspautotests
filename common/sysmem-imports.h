extern int sceKernelSetCompiledSdkVersion(u32 ver);
extern int sceKernelSetCompiledSdkVersion370(u32 ver);
extern int sceKernelSetCompiledSdkVersion380_390(u32 ver);
extern int sceKernelSetCompiledSdkVersion395(u32 ver);
extern int sceKernelSetCompiledSdkVersion401_402(u32 ver);
extern int sceKernelSetCompiledSdkVersion500_505(u32 ver);
extern int sceKernelSetCompiledSdkVersion507(u32 ver);
extern int sceKernelSetCompiledSdkVersion600_602(u32 ver);
extern int sceKernelSetCompiledSdkVersion603_605(u32 ver);
extern int sceKernelSetCompiledSdkVersion606(u32 ver);
extern u32 sceKernelGetCompiledSdkVersion();