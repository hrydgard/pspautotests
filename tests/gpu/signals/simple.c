#include <common.h>

#include <stdlib.h>
#include <stdio.h>
#include <pspgu.h>
#include <pspgum.h>

#include "sysmem-imports.h"

extern int sceGeContinue();
extern int sceGeBreak(int breakType);

static unsigned int __attribute__((aligned(16))) list[262144];

char buffer[65535];
char *bpos = &buffer[0];

#define BUF_WIDTH (512)
#define SCR_WIDTH (480)
#define SCR_HEIGHT (272)
#define PIXEL_SIZE (4) /* change this if you change to another screenmode */
#define FRAME_SIZE (BUF_WIDTH * SCR_HEIGHT * PIXEL_SIZE)
#define ZBUF_SIZE (BUF_WIDTH * SCR_HEIGHT * 2) /* zbuffer seems to be 16-bit? */

unsigned int __attribute__((aligned(16))) dlist1[] = {
	0x00000000, // 0x00 NOP
	0x0E010000, // 0x01 SIGNAL + WAIT
	0x0C000000, // 0x02 END
	0x00000000, // 0x03 NOP
	0x00000000, // 0x04 NOP
	0x00000000, // 0x05 NOP
	0x00000000, // 0x06 NOP
	0x00000000, // 0x07 NOP
	0x00000000, // 0x08 NOP
	0x00000000, // 0x09 NOP
	0x0F000000, // 0x0A FINISH
	0x0C000000, // 0x0B END
	0x00000000, // 0x0C NOP
	0x00000000, // 0x0D NOP
	0x0F000000, // 0x0E FINISH
	0x0C000000, // 0x0F END
};

int dlist1id, dlist2id;
int cbid1, cbid2;

char* status_str(int status) {
	if(status < 0 || status > 4)
		return "INVALID";

	char* str[] = {
		"DONE",
		"QUEUED",
		"DRAWING",
		"STALL",
		"PAUSED",
	};
	return str[status];
}

inline void breakInfo(const char *format, ...) {
	bpos += sprintf(bpos, "  BREAK \t");

	int result = sceGeBreak(1);
	bpos += sprintf(bpos, "  %-8s\t%08x", "", result);

	va_list args;
	va_start(args, format);
	bpos += vsprintf(bpos, format, args);
	va_end(args);
}

inline void syncInfo(const char *format, ...) {
	bpos += sprintf(bpos, "  WAIT  \t");

	int drawsync = sceGeDrawSync(0);
	bpos += sprintf(bpos, "%x %-8s\t", drawsync, status_str(drawsync));

	va_list args;
	va_start(args, format);
	bpos += vsprintf(bpos, format, args);
	va_end(args);
}

inline void listInfo(int n, const char *format, ...) {
	if (n != 0) {
		bpos += sprintf(bpos, "  List %d\t", n);
	} else {
		bpos += sprintf(bpos, "        \t");
	}

	int drawsync = sceGeDrawSync(1);
	bpos += sprintf(bpos, "%x %-8s\t", drawsync, status_str(drawsync));

	va_list args;
	va_start(args, format);
	bpos += vsprintf(bpos, format, args);
	va_end(args);
}

inline void listInfoNosync(int n, const char *format, ...) {
	if (n != 0) {
		bpos += sprintf(bpos, "  List %d\t", n);
	} else {
		bpos += sprintf(bpos, "        \t");
	}

	bpos += sprintf(bpos, "  %-8s\t", "");

	va_list args;
	va_start(args, format);
	bpos += vsprintf(bpos, format, args);
	va_end(args);
}

int ge_signal(int value, void* arg) {
	unsigned int *addr;
    asm("sw $a2, %0" : "=m"(addr));

	unsigned int pos = addr == NULL ? 0xFF : (unsigned int) (addr - dlist1);
	listInfo((int) arg + 1, "Signal(%x, list+%02X)\n", (unsigned int) value, pos);
	return 0;
}

int ge_finish(int value, void* arg) {
	unsigned int *addr;
    asm("sw $a2, %0" : "=m"(addr));

	unsigned int pos = addr == NULL ? 0xFF : (unsigned int) (addr - dlist1);
	listInfo((int) arg + 1, "Finish(%x, list+%02X)\n", (unsigned int) value, pos);
	return 0;
}

void init() {
	sceKernelDcacheWritebackAll();

	sceGuInit();
	sceGuStart(GU_DIRECT, list);

	sceGuDrawBuffer (GU_PSM_8888, (void*)FRAME_SIZE, BUF_WIDTH);
	sceGuDepthBuffer((void *)(FRAME_SIZE * 2), BUF_WIDTH);
	sceGuOffset     (2048 - (SCR_WIDTH / 2),2048 - (SCR_HEIGHT / 2));
	sceGuViewport   (2048, 2048, SCR_WIDTH, SCR_HEIGHT);
	sceGuDepthRange (0xc350, 0x2710);
	sceGuScissor    (0, 0, SCR_WIDTH, SCR_HEIGHT);
	sceGuFinish     ();
	sceGuSync       (GU_SYNC_FINISH, GU_SYNC_WHAT_DONE);

	PspGeCallbackData cbdata;
	cbdata.signal_func = (PspGeCallback) ge_signal;
	cbdata.signal_arg  = NULL;
	cbdata.finish_func = (PspGeCallback) ge_finish;
	cbdata.finish_arg  = NULL;
	cbid1 = sceGeSetCallback(&cbdata);

	PspGeCallbackData cbdata2;
	cbdata2.signal_func = (PspGeCallback) ge_signal;
	cbdata2.signal_arg  = (void*) 1;
	cbdata2.finish_func = (PspGeCallback) ge_finish;
	cbdata2.finish_arg  = (void*) 1;
	cbid2 = sceGeSetCallback(&cbdata2);
}

enum {
	TEST_USE_DRAWSYNC = 0x00,
	TEST_USE_BREAK = 0x01,
	TEST_STALL_LATE = 0x02,
};

void testGeCallbacks(int method) {
	int listsync, result;
	sceKernelDcacheWritebackAll();

	bpos += sprintf(bpos, "  LIST #\tDRAWSTATE\tINFO\n");

	dlist1id = sceGeListEnQueue(dlist1, 0, cbid1, 0);
	listInfoNosync(1, "Enqueued without stall...\n");
	listsync = sceGeListSync(dlist1id, 1);
	listInfo(1, "Sync %x %s\n", listsync, status_str(listsync));

	result = sceGeContinue();
	listsync = sceGeListSync(dlist1id, 1);
	listInfo(1, "Sync %x %s after continue (%08x)\n", listsync, status_str(listsync), result);

	dlist2id = sceGeListEnQueue(dlist1, method & TEST_STALL_LATE ? dlist1 + 0x0D : dlist1 + 0x05, cbid2, 0);
	listInfoNosync(2, "Enqueued with %s...\n", method & TEST_STALL_LATE ? "late stall" : "stall");
	listsync = sceGeListSync(dlist2id, 1);
	listInfo(2, "Sync %x %s\n", listsync, status_str(listsync));

	result = sceGeContinue();
	listsync = sceGeListSync(dlist2id, 1);
	listInfo(2, "Sync %x %s after continue (%08x)\n", listsync, status_str(listsync), result);

	result = sceGeListUpdateStallAddr(dlist2id, 0);
	listsync = sceGeListSync(dlist2id, 1);
	listInfo(2, "Sync %x %s after unstall (%08x)\n", listsync, status_str(listsync), result);

	printf("%s", buffer);
	bpos = buffer;
	bpos[0] = '\0';

	if (method & TEST_USE_BREAK) {
		breakInfo("\n");
	} else {
		syncInfo("\n");
	}

	printf("%s", buffer);
	bpos = buffer;
	bpos[0] = '\0';
}

enum
{
	GE_SIG_SUSPEND = 0x01,
	GE_SIG_CONTINUE = 0x02,
	GE_SIG_PAUSE = 0x03,
	GE_SIG_SYNC = 0x08,

	GE_SIG_BP1 = 0xF0,
	GE_SIG_BP2 = 0xFF,
};

#define MAKE_GE_END(type, value) (0x0C000000 | ((type & 0xFF) << 16) | (value & 0xFFFF))
#define MAKE_GE_SIGNAL(type, value) (0x0E000000 | ((type & 0xFF) << 16) | (value & 0xFFFF))
#define MAKE_GE_FINISH(type, value) (0x0F000000 | ((type & 0xFF) << 16) | (value & 0xFFFF))

void testSignalTypes() {
	// All invalid callbacks should work this way? (0x04, 0xFE, etc.)
	printf("\nUnknown (0x%02x):\n", 0);
	dlist1[1] = MAKE_GE_SIGNAL(0, 0x1234);
	dlist1[2] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_BREAK | TEST_STALL_LATE);

	// SUSPEND list while in handler
	printf("\nSignal handler + wait (0x%02x):\n", GE_SIG_SUSPEND);
	dlist1[0x01] = MAKE_GE_SIGNAL(GE_SIG_SUSPEND, 0x1234);
	dlist1[0x02] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_DRAWSYNC);

	// CONTINUE list keeps getting processed
	printf("\nSignal handler only (0x%02x):\n", GE_SIG_CONTINUE);
	dlist1[0x01] = MAKE_GE_SIGNAL(GE_SIG_CONTINUE, 0x1234);
	dlist1[0x02] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_DRAWSYNC);

	// PAUSE list and call handler at FINISH CMD, but can resume
	printf("\nSignal handler + pause (0x03):\n");
	dlist1[0x01] = MAKE_GE_SIGNAL(GE_SIG_PAUSE, 0x1234);
	dlist1[0x02] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_DRAWSYNC | TEST_STALL_LATE);

	// SYNC doesn't call handler but syncs and skips
	printf("\nSync + continue (0x%02x):\n", GE_SIG_SYNC);
	dlist1[0x01] = MAKE_GE_SIGNAL(GE_SIG_SYNC, 0x1234);
	dlist1[0x02] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_DRAWSYNC | TEST_STALL_LATE);

	// Not included in this test - 0x10 - 0x16 (call/jump/ret/etc.)
	// Not included in this test - 0x20 - 0x38 (tex/clut/etc.)

	// BREAKPOINT may not do anything?
	printf("\nBreakpoint #1 (0x%02x):\n", GE_SIG_BP1);
	dlist1[0x01] = MAKE_GE_SIGNAL(GE_SIG_BP1, 0x1234);
	dlist1[0x02] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_DRAWSYNC);

	printf("\nUnknown (0x%02x):\n", 0xEE);
	dlist1[1] = MAKE_GE_SIGNAL(0xEE, 0x1234);
	dlist1[2] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_BREAK | TEST_STALL_LATE);

	printf("\nBreakpoint #2 (0x%02x):\n", GE_SIG_BP2);
	dlist1[0x01] = MAKE_GE_SIGNAL(GE_SIG_BP2, 0x1234);
	dlist1[0x02] = MAKE_GE_END(0, 0);
	testGeCallbacks(TEST_USE_DRAWSYNC);
}

int main(int argc, char *argv[]) {
	printf("Using default SDK version:\n");
	init();
	testSignalTypes();

	printf("\n\nUsing 6.60 SDK version:\n");
	sceKernelSetCompiledSdkVersion(0x6060010);
	init();
	testSignalTypes();

	return 0;
}
