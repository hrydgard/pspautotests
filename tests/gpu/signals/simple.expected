Using default SDK version:

Unknown (0x00):
  LIST #	DRAWSTATE	INFO
  List 1	          	Enqueued without stall...
  List 1	2 DRAWING 	Sync 2 DRAWING
  List 1	2 DRAWING 	Sync 2 DRAWING after continue (ffffffff)
  List 2	          	Enqueued with late stall...
  List 2	2 DRAWING 	Sync 1 QUEUED
  List 2	2 DRAWING 	Sync 1 QUEUED after continue (ffffffff)
  List 2	2 DRAWING 	Sync 1 QUEUED after unstall (00000000)
  BREAK 	          	00000000

Signal handler + wait (0x01):
  LIST #	DRAWSTATE	INFO
  List 1	2 DRAWING 	Signal(1234, list+FF)
  List 1	0 DONE    	Finish(0, list+FF)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	2 DRAWING 	Signal(1234, list+FF)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (ffffffff)
  List 2	0 DONE    	Finish(0, list+FF)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Signal handler only (0x02):
  LIST #	DRAWSTATE	INFO
  List 1	2 DRAWING 	Signal(1234, list+FF)
  List 1	0 DONE    	Finish(0, list+FF)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	3 STALL   	Signal(1234, list+FF)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (ffffffff)
  List 2	0 DONE    	Finish(0, list+FF)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Signal handler + pause (0x03):
  LIST #	DRAWSTATE	INFO
  List 1	2 DRAWING 	Signal(1234, list+FF)
  List 1	          	Enqueued without stall...
  List 1	2 DRAWING 	Sync 4 PAUSED
  List 1	0 DONE    	Finish(0, list+FF)
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	2 DRAWING 	Signal(1234, list+FF)
  List 2	          	Enqueued with late stall...
  List 2	2 DRAWING 	Sync 4 PAUSED
  List 2	3 STALL   	Sync 3 STALL after continue (00000000)
  List 2	0 DONE    	Finish(0, list+FF)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Sync + continue (0x08):
  LIST #	DRAWSTATE	INFO
  List 1	0 DONE    	Finish(0, list+FF)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	          	Enqueued with late stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (ffffffff)
  List 2	0 DONE    	Finish(0, list+FF)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Breakpoint #1 (0xf0):
  LIST #	DRAWSTATE	INFO
  List 1	0 DONE    	Finish(0, list+FF)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (ffffffff)
  List 2	0 DONE    	Finish(0, list+FF)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Unknown (0xee):
  LIST #	DRAWSTATE	INFO
  List 1	          	Enqueued without stall...
  List 1	2 DRAWING 	Sync 2 DRAWING
  List 1	2 DRAWING 	Sync 2 DRAWING after continue (ffffffff)
  List 2	          	Enqueued with late stall...
  List 2	2 DRAWING 	Sync 1 QUEUED
  List 2	2 DRAWING 	Sync 1 QUEUED after continue (ffffffff)
  List 2	2 DRAWING 	Sync 1 QUEUED after unstall (00000000)
  BREAK 	          	00000000

Breakpoint #2 (0xff):
  LIST #	DRAWSTATE	INFO
  List 1	0 DONE    	Finish(0, list+FF)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (ffffffff)
  List 2	0 DONE    	Finish(0, list+FF)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	


Using 6.60 SDK version:

Unknown (0x00):
  LIST #	DRAWSTATE	INFO
  List 1	          	Enqueued without stall...
  List 1	2 DRAWING 	Sync 2 DRAWING
  List 1	2 DRAWING 	Sync 2 DRAWING after continue (80000020)
  List 2	          	Enqueued with late stall...
  List 2	2 DRAWING 	Sync 80000100 INVALID
  List 2	2 DRAWING 	Sync 80000100 INVALID after continue (80000020)
  List 2	2 DRAWING 	Sync 80000100 INVALID after unstall (80000100)
  BREAK 	          	00000000

Signal handler + wait (0x01):
  LIST #	DRAWSTATE	INFO
  List 1	2 DRAWING 	Signal(1234, list+03)
  List 1	0 DONE    	Finish(0, list+0C)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	2 DRAWING 	Signal(1234, list+03)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (80000020)
  List 2	0 DONE    	Finish(0, list+0C)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Signal handler only (0x02):
  LIST #	DRAWSTATE	INFO
  List 1	2 DRAWING 	Signal(1234, list+03)
  List 1	0 DONE    	Finish(0, list+0C)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	3 STALL   	Signal(1234, list+03)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (80000020)
  List 2	0 DONE    	Finish(0, list+0C)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Signal handler + pause (0x03):
  LIST #	DRAWSTATE	INFO
  List 1	2 DRAWING 	Signal(1234, list+0C)
  List 1	          	Enqueued without stall...
  List 1	2 DRAWING 	Sync 4 PAUSED
  List 1	0 DONE    	Finish(0, list+10)
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	2 DRAWING 	Signal(1234, list+0C)
  List 2	          	Enqueued with late stall...
  List 2	2 DRAWING 	Sync 4 PAUSED
  List 2	3 STALL   	Sync 3 STALL after continue (00000000)
  List 2	0 DONE    	Finish(0, list+10)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Sync + continue (0x08):
  LIST #	DRAWSTATE	INFO
  List 1	0 DONE    	Finish(0, list+10)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	          	Enqueued with late stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (80000020)
  List 2	0 DONE    	Finish(0, list+10)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Breakpoint #1 (0xf0):
  LIST #	DRAWSTATE	INFO
  List 1	0 DONE    	Finish(0, list+0C)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (80000020)
  List 2	0 DONE    	Finish(0, list+0C)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	

Unknown (0xee):
  LIST #	DRAWSTATE	INFO
  List 1	          	Enqueued without stall...
  List 1	2 DRAWING 	Sync 2 DRAWING
  List 1	2 DRAWING 	Sync 2 DRAWING after continue (80000020)
  List 2	          	Enqueued with late stall...
  List 2	2 DRAWING 	Sync 80000100 INVALID
  List 2	2 DRAWING 	Sync 80000100 INVALID after continue (80000020)
  List 2	2 DRAWING 	Sync 80000100 INVALID after unstall (80000100)
  BREAK 	          	00000000

Breakpoint #2 (0xff):
  LIST #	DRAWSTATE	INFO
  List 1	0 DONE    	Finish(0, list+0C)
  List 1	          	Enqueued without stall...
  List 1	0 DONE    	Sync 0 DONE
  List 1	0 DONE    	Sync 0 DONE after continue (00000000)
  List 2	          	Enqueued with stall...
  List 2	3 STALL   	Sync 3 STALL
  List 2	3 STALL   	Sync 3 STALL after continue (80000020)
  List 2	0 DONE    	Finish(0, list+0C)
  List 2	0 DONE    	Sync 0 DONE after unstall (00000000)
  WAIT  	0 DONE    	
